libcaf_core/caf/action.hpp

```diff
@@ -156,8 +162,9 @@ struct default_action_impl : detail::atomic_ref_counted, action::impl {
   void run() override {
     if (state_.load() == action::state::scheduled) {
       f_();
-      if constexpr (IsSingleShot)
-        state_ = action::state::disposed;
+      if constexpr (IsSingleShot) {
+        dispose();
+      }
```

Comment:
This code can now run into UB, e.g. segfault or do other weird stuff.

- Thread 1 is calling run() and passes line 164
- Thread 2 is calling dispose() and runs the destructor of f_
- Thread 1 calls f_(): boom; accessing a destructed object

There's an easy fix for single-shot actions, but multi-shot actions are tricky to implement lock-free. The multi-shot type is used much more widely in the code base, so we need a good solution here.